registry/storage/catalog.go

```diff
@@ -96,22 +96,11 @@ func (reg *registry) Enumerate(ctx context.Context, ingester func(string) error)
 
 }
 
-func pathGreaterThan(pathX, pathY string) (b bool) {
-	splitPathX := strings.SplitN(pathX, "/", 2)
-	splitPathY := strings.SplitN(pathY, "/", 2)
-
-	if splitPathX[0] == splitPathY[0] {
-		if len(splitPathX) == 1 && len(splitPathY) == 1 {
-			return false
-		} else if len(splitPathX) == 1 && len(splitPathY) != 1 {
-			return false
-		} else if len(splitPathX) != 1 && len(splitPathY) == 1 {
-			return true
-		}
-
-		return pathGreaterThan(splitPathX[1], splitPathY[1])
-
-	}
-
-	return splitPathX[0] > splitPathY[0]
+// lessPath returns true if one path a is less than path b.
+//
+// A component-wise comparison is done, rather than the lexical comparison of
+// strings.
+func lessPath(a, b string) bool {
+	// we provide this behavior by making separator always sort first.
+	return strings.Replace(a, "/", "\x00", -1) < strings.Replace(b, "/", "\x00", -1)
```

Comment:
If garbage is a concern, it may be better to avoid mutating strings. You could have a simple comparison function that iterates character-by-character and compares the first differing component.